In [2]:
import pandas as pd
import numpy as np
from IPython.display import display, HTML
from yf_utils import _2_split_train_val_test, _3_random_slices, _4_lookback_slices
from yf_utils import _5_perf_ranks, _6_grp_tuples_sort_sum
from myUtils import pickle_load, pickle_dump

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 30)
pd.set_option('display.max_colwidth', 16)
pd.set_option('display.width', 790)

path_dir = "C:/Users/ping/MyDrive/stocks/yfinance/"
path_data_dump = path_dir + "VSCode_dump/"
fp_df_close_clean = 'df_close_clean'

# fp_df_eval_results = 'df_eval_results'
# fp_df_eval_sym_freq_results = 'df_eval_sym_freq_results'
fp_df_eval_sym_freq_results_test = 'df_eval_sym_freq_results_test'

df_close_clean = pickle_load(path_data_dump, fp_df_close_clean)

In [3]:
# Split df_close_clean into training (df_train), validation (df_val) and test (df_test) set.
# The default split is 0.7, 0.2, 0.1 respectively.

###################################  USE df_test as df_train
df_train, df_val, df_test = _2_split_train_val_test(df_close_clean)
df_train = df_test.copy()
###################################  USE df_test as df_train

len_df_train = len(df_train)
len_df_val = len(df_val)
len_df_test = len(df_test)
print(f'len_df_train: {len_df_train}, len_df_val: {len_df_val}, len_df_test: {len_df_test} ')

len_df_train: 150, len_df_val: 300, len_df_test: 150 


In [4]:
print(f'df_train.index: {df_train.index[0]}, {df_train.index[-1]}')
print(f'df_val.index:   {df_val.index[0]}, {df_val.index[-1]}')

df_train.index: 2022-07-22 00:00:00, 2023-02-24 00:00:00
df_val.index:   2021-05-13 00:00:00, 2022-07-21 00:00:00


In [5]:
col_add0 = ['n_samples', 'days_lookbacks', 'days_eval', 'n_top_syms', 'syms_start', 'syms_end', 'sym_freq_cnt']
col_add1 = ['grp(CAGR)_mean',         'grp(CAGR)_std',         'grp(CAGR)_mean/std']
col_add2 = ['grp(CAGR/UI)_mean',      'grp(CAGR/UI)_std',      'grp(CAGR/UI)_mean/std']
col_add3 = ['grp(CAGR/retnStd)_mean', 'grp(CAGR/retnStd)_std', 'grp(CAGR/retnStd)_mean/std']
col_add4 = ['grp(retnStd/UI)_mean',   'grp(retnStd/UI)_std',   'grp(retnStd/UI)_mean/std']
col_add5 = ['SPY_CAGR', 'SPY_CAGR/UI', 'SPY_CAGR/retnStd', 'SPY_retnStd/UI']

col_add_total = col_add0 + col_add1 + col_add2 + col_add3 + col_add4 + col_add5
print(f'col_add_total:\n{col_add_total}, total columns: {len(col_add_total)}')

col_add_total:
['n_samples', 'days_lookbacks', 'days_eval', 'n_top_syms', 'syms_start', 'syms_end', 'sym_freq_cnt', 'grp(CAGR)_mean', 'grp(CAGR)_std', 'grp(CAGR)_mean/std', 'grp(CAGR/UI)_mean', 'grp(CAGR/UI)_std', 'grp(CAGR/UI)_mean/std', 'grp(CAGR/retnStd)_mean', 'grp(CAGR/retnStd)_std', 'grp(CAGR/retnStd)_mean/std', 'grp(retnStd/UI)_mean', 'grp(retnStd/UI)_std', 'grp(retnStd/UI)_mean/std', 'SPY_CAGR', 'SPY_CAGR/UI', 'SPY_CAGR/retnStd', 'SPY_retnStd/UI'], total columns: 23


In [6]:
# # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_add_total)
# pickle_dump(df, path_data_dump, fp_df_eval_sym_freq_results_test)
# df

Empty DataFrame
Columns: [n_samples, days_lookbacks, days_eval, n_top_syms, syms_start, syms_end, sym_freq_cnt, grp(CAGR)_mean, grp(CAGR)_std, grp(CAGR)_mean/std, grp(CAGR/UI)_mean, grp(CAGR/UI)_std, grp(CAGR/UI)_mean/std, grp(CAGR/retnStd)_mean, grp(CAGR/retnStd)_std, grp(CAGR/retnStd)_mean/std, grp(retnStd/UI)_mean, grp(retnStd/UI)_std, grp(retnStd/UI)_mean/std, SPY_CAGR, SPY_CAGR/UI, SPY_CAGR/retnStd, SPY_retnStd/UI]
Index: []

In [7]:
# if store_results:  # record results to df
# my_cols = ['n_samples', 'days_lookbacks', 'days_eval', 'n_top_syms', 'syms_start', 'syms_end', 'grp(CAGR/UI)_mean', 'grp(CAGR/UI)_std', 'grp(CAGR/UI)_mean/std', 'SPY_CAGR/UI']
# df_eval_results = pickle_load(path_data_dump, fp_df_eval_results)
df_eval_sym_freq_results_test = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_test)
df_eval_sym_freq_results_test

Empty DataFrame
Columns: [n_samples, days_lookbacks, days_eval, n_top_syms, syms_start, syms_end, sym_freq_cnt, grp(CAGR)_mean, grp(CAGR)_std, grp(CAGR)_mean/std, grp(CAGR/UI)_mean, grp(CAGR/UI)_std, grp(CAGR/UI)_mean/std, grp(CAGR/retnStd)_mean, grp(CAGR/retnStd)_std, grp(CAGR/retnStd)_mean/std, grp(retnStd/UI)_mean, grp(retnStd/UI)_std, grp(retnStd/UI)_mean/std, SPY_CAGR, SPY_CAGR/UI, SPY_CAGR/retnStd, SPY_retnStd/UI]
Index: []

In [476]:
# verbose = True  # True prints more output
verbose = False  # True prints more output

# write run results to df_eval_results
# store_results = False
store_results = True

# number of max lookback tuples to create for iloc start_train:end_train:end_eval
# i.e. number of grp_top_set_syms_n_freq and grp_top_set_syms 
# n_samples = 114  
n_samples = 57  
# n_samples = 1

# for training, the number of days to lookback from iloc max-lookback end_train
days_lookbacks = [15, 30, 60, 120]
days_lookbacks = [30, 60, 120]
days_lookbacks = [60, 120]
days_lookbacks = [120]
days_lookbacks = [60]
days_lookbacks = [30]
days_lookbacks = [15]
days_lookbacks = [15, 30]
days_lookbacks = [15, 30, 60]


days_lookbacks.sort()

# number of days from end_train are used to evaluate effectiveness of the training
# days_eval = 10
days_eval = 6
days_eval = 5
days_eval = 4
# days_eval = 3
# days_eval = 2  


# number of the most-common symbols from days_lookbacks' performance rankings to keep
n_top_syms = 20  

syms_start = 0  #  start index of n_top_syms for evaluation
# syms_start = 1  #  start index of n_top_syms for evaluation

# syms_end = n_top_syms  #  end index of n_top_syms for evaluation
# syms_end = 1  #  end index of n_top_syms for evaluation
# syms_end = 2  #  end index of n_top_syms for evaluation
# syms_end = 3  #  end index of n_top_syms for evaluation
# syms_end = 4  #  end index of n_top_syms for evaluation
# syms_end = 5  #  end index of n_top_syms for evaluation
# syms_end = 6  #  end index of n_top_syms for evaluation
syms_end = 10  #  end index of n_top_syms for evaluation

create a sets of iloc lookback slices (start_train:end_train:end_eval), where  
end_train - start_train = days_lookback  
end_eval - end_train = days_eval  
for example,  
if given:  
 n_samples = 2  
 days_lookbacks = [30, 60, 120]  
 days_eval = 10  
a possible result is:  
 max_lookback_slices:  
 [(150, 270, 280), (5, 125, 135)]  
 where 270-150=125-5=max(days_lookbacks), 280-270=135-125=days_eval  
 sets_lookback_slices:  
 [[(240, 270, 280), (210, 270, 280), (150, 270, 280)], [(95, 125, 135), (65, 125, 135), (5, 125, 135)]]  
  where in a set, 270-240=days_lookbacks[0], 270-210=days_lookbacks[1], 270-150=days_lookbacks[2]  
  and 270, i.e. end_train, is constant for the set  

In [477]:
# return n_samples slices
max_lookback_slices = _3_random_slices(len_df_train, n_samples=n_samples, days_lookback=max(days_lookbacks), days_eval=days_eval)
# return n_samples * len(days_lookbacks) slices
sets_lookback_slices = _4_lookback_slices(max_slices=max_lookback_slices, days_lookbacks=days_lookbacks, verbose=False)

if verbose:
  print(f'max_lookback_slices:\n{max_lookback_slices}')
  print(f'sets_lookback_slices:\n{sets_lookback_slices}')

Generate lists, n_samples long, of the highest performance ranked symbols. The performance metrics are: CAGR/UI, CAGR/retnStd, retnStd/UI. n_top_syms of the best performing symbols from each metric are combined. The symbols are sorted by their number of appearances in the combined pool, and are placed in a list. A slice of the best performing symbols is selected by syms_start:syms_end, i.e. top_set_syms_n_freq[syms_start:syms_end].     

The performance metrics are calculated based on slices in sets_lookback_slices.  The first two numbers are ilocs for training. The last two numbers are ilocs for evaluation.

In [478]:
grp_top_set_syms_n_freq = []  # list of lists of top_set_symbols_n_freq, there are n_samples lists in list
grp_top_set_syms = []  # list of lists of top_set_symbols, there are n_samples lists in list
# loop thru lists of tuples of start_train:end_train:end_eval, i.e.
#  [[(887, 917, 927), (857, 917, 927), (797, 917, 927)],
#  [(483, 513, 523), (453, 513, 523), (393, 513, 523)]]
for lb_slices in sets_lookback_slices:
  grp_most_common_syms = []  
  for lb_slice in lb_slices:  # lb_slice, e.g. (246, 276, 286)
    start_train = lb_slice[0]
    end_train = lb_slice[1]
    start_eval = end_train
    end_eval = lb_slice[2]
    lookback = end_train - start_train
    d_eval = end_eval - start_eval

    _df = df_train.iloc[start_train:end_train]
    date_start_train = _df.index[0].strftime('%Y-%m-%d')
    date_end_train = _df.index[-1].strftime('%Y-%m-%d')

    if verbose:
      print(f'days lookback:    {lookback}')
      print(f'lb_slices:        {lb_slices}')
      print(f'lb_slice:         {lb_slice}')
      print(f'days eval:        {d_eval}')    
      print(f'iloc_start_train: {start_train}')
      print(f'iloc_end_train:   {end_train}')
      print(f'date_start_train: {date_start_train}')
      print(f'date_end_train:   {date_end_train}')


    perf_ranks, most_common_syms = _5_perf_ranks(_df, n_top_syms=n_top_syms)
    grp_most_common_syms.append(most_common_syms)
    
    if verbose:    
      # 1 lookback of r_CAGR/UI, r_CAGR/retnStd, r_retnStd/UI
      print(f'perf_ranks: {perf_ranks}')  
      # most common symbols of perf_ranks 
      print(f'most_common_syms: {most_common_syms}')     
      # grp_perf_ranks[lookback] = perf_ranks
      print(f'+++ finish lookback slice {lookback} +++\n')

  if verbose:
    print(f'grp_most_common_syms: {grp_most_common_syms}')
    # grp_most_common_syms a is list of lists of tuples of 
    #  the most-common-symbols symbol:frequency cumulated from
    #  each days_lookback  
    print(f'**** finish lookback slices {lb_slices} ****\n')

  # flatten list of lists of (symbol:frequency)
  flat_grp_most_common_syms = [val for sublist in grp_most_common_syms for val in sublist]
  # group symbols from set of days_lookbacks (i.e. lb_slices) and sum frequency of the symbols
  set_most_common_syms = _6_grp_tuples_sort_sum(flat_grp_most_common_syms, reverse=True)
  # get the top few most-frequent symbol:frequency pairs
  top_set_syms_n_freq = set_most_common_syms[0:n_top_syms]
  # get symbols from top_set_syms_n_freq

###################################  
  # top_set_syms = [i[0] for i in top_set_syms_n_freq]
  top_set_syms = [i[0] for i in top_set_syms_n_freq[syms_start:syms_end]]  
###################################  
  
  grp_top_set_syms_n_freq.append(top_set_syms_n_freq)
  grp_top_set_syms.append(top_set_syms)

  if verbose:  
    print(f'top {n_top_syms} ranked symbols and frequency from set {lb_slices}:\n{top_set_syms_n_freq}')
    print(f'top {n_top_syms} ranked symbols from set {lb_slices}:\n{top_set_syms}')  
    print(f'===== finish top {n_top_syms} ranked symbols from days_lookback set {lb_slices} =====\n\n')

In [479]:
def top_set_sym_freq_cnt(top_set_syms_n_freq):
    sym_freq_cnt_6 = []
    sym_freq_cnt_5 = []
    sym_freq_cnt_4 = []
    sym_freq_cnt_3 = []
    sym_freq_cnt_2 = []

    for sym_n_freq in top_set_syms_n_freq:
        _sym = sym_n_freq[0]
        _freq = sym_n_freq[1]
        # print(_sym, _freq)
        if _freq == 6:
            sym_freq_cnt_6.append(_sym)
        elif _freq == 5:
            sym_freq_cnt_5.append(_sym)
        elif _freq == 4:
            sym_freq_cnt_4.append(_sym)
        elif _freq == 3:
            sym_freq_cnt_3.append(_sym)          
        else:
            sym_freq_cnt_2.append(_sym)

    l_sym_freq_cnt = []
    l_sym_freq_cnt.append(sym_freq_cnt_6)
    l_sym_freq_cnt.append(sym_freq_cnt_5)
    l_sym_freq_cnt.append(sym_freq_cnt_4)
    l_sym_freq_cnt.append(sym_freq_cnt_3)    
    l_sym_freq_cnt.append(sym_freq_cnt_2)    

    # if verbose:
    #     print(f'sym_freq_cnt_6: {sym_freq_cnt_6}')
    #     print(f'sym_freq_cnt_5: {sym_freq_cnt_5}')
    #     print(f'sym_freq_cnt_4: {sym_freq_cnt_4}')
    #     print(f'sym_freq_cnt_3: {sym_freq_cnt_3}')
    #     print(f'sym_freq_cnt_2: {sym_freq_cnt_2}')

    # return sym_freq_cnt_6, sym_freq_cnt_5, sym_freq_cnt_4, sym_freq_cnt_3, sym_freq_cnt_2
    return l_sym_freq_cnt    

In [480]:
for top_set_syms_n_freq in grp_top_set_syms_n_freq:
  l_sym_freq_cnt = top_set_sym_freq_cnt(top_set_syms_n_freq)
  print(f'sym_freq_6: {l_sym_freq_cnt[0]}')
  print(f'sym_freq_5: {l_sym_freq_cnt[1]}')
  print(f'sym_freq_4: {l_sym_freq_cnt[2]}')
  print(f'sym_freq_3: {l_sym_freq_cnt[3]}')
  print(f'sym_freq_2: {l_sym_freq_cnt[4]}\n')  

sym_freq_6: ['AXON', 'GILD']
sym_freq_5: ['GBX', 'LOCO', 'RMBS', 'SLB', 'STRL']
sym_freq_4: ['ASC', 'ELF', 'EME', 'SANM']
sym_freq_3: ['ATEN', 'CAT', 'CCRN', 'FTI', 'HEES', 'HOG', 'NTNX', 'SKT', 'TWNK']
sym_freq_2: []

sym_freq_6: ['BVH']
sym_freq_5: ['BGNE', 'COLL', 'HZNP', 'NEOG']
sym_freq_4: ['BABA', 'EDU', 'EXAS', 'NATI', 'NSTG', 'NTES', 'WB']
sym_freq_3: ['ACIW', 'ACLS', 'ALGN', 'AU', 'BMA', 'BURL', 'CROX', 'DENN']
sym_freq_2: []

sym_freq_6: ['ELF', 'GILD']
sym_freq_5: ['AMKR', 'EME', 'RMBS']
sym_freq_4: ['ASC', 'DD', 'LOCO']
sym_freq_3: ['BIIB', 'CAT', 'DENN', 'FTI', 'HFWA', 'LVS', 'MODN', 'NTNX', 'NYT', 'OSBC', 'PKX']
sym_freq_2: ['AXON']

sym_freq_6: ['ZTO']
sym_freq_5: ['AXSM', 'BIDU', 'GDS']
sym_freq_4: ['AU', 'BVH', 'CTRN', 'PUK', 'WB']
sym_freq_3: ['ACLS', 'AER', 'ATKR', 'BERY', 'BURL', 'COUP', 'ELF', 'GILD']
sym_freq_2: ['BBW', 'HZNP', 'NVO']

sym_freq_6: ['HLIT']
sym_freq_5: ['ASC', 'CAR', 'CATO', 'LOCO', 'PBF', 'SIGI', 'SLB']
sym_freq_4: ['BBW', 'ISEE', 'STLD', 'XPRO']


In [481]:
from myUtils import symb_perf_stats_vectorized_v8

print('z_grp_top_set_syms_n_freq:')
z_grp_top_set_syms_n_freq = zip(max_lookback_slices, grp_top_set_syms_n_freq)

for i, (_lookback_slice, _top_set_syms_n_freq) in enumerate(z_grp_top_set_syms_n_freq):
  start_train = _lookback_slice[0]
  end_train = _lookback_slice[1]
  start_eval = end_train
  end_eval = _lookback_slice[2]

  print(f'{i + 1 } of {n_samples} max_lookback_slice')
  print(f'max_lookback_slice: {_lookback_slice}')
  # dates correspond to max_lookback_slice
  d_start_train = df_train.index[start_train].strftime('%Y-%m-%d')
  d_end_train = df_train.index[end_train].strftime('%Y-%m-%d')
  d_start_eval = d_end_train
  d_end_eval = df_train.index[end_eval].strftime('%Y-%m-%d')
  d_df_eval_start = d_end_train
  d_df_eval_end = df_train.index[end_eval - 1].strftime('%Y-%m-%d')  
  print(f'max lookback dates: {d_start_train}, {d_end_train}, {d_end_eval}')
  print(f'df_eval dates (inclusive): {d_df_eval_start} - {d_df_eval_end}')    
  print(f'top_set_syms_n_freq: {_top_set_syms_n_freq}\n')

  l_sym_freq_cnt = top_set_sym_freq_cnt(_top_set_syms_n_freq)
  print(f'sym_freq_6: {l_sym_freq_cnt[0]}')
  print(f'sym_freq_5: {l_sym_freq_cnt[1]}')
  print(f'sym_freq_4: {l_sym_freq_cnt[2]}')
  print(f'sym_freq_3: {l_sym_freq_cnt[3]}\n')

  _sym_idx = ['SPY']
  df_SPY = df_train[start_eval:end_eval][_sym_idx]

  (
    _symbols,
    _period_yr,
    _retn,
    _DD,
    _UI,
    _MDD,
    _retnMean,
    _retnStd,
    _retnStd_div_UI,
    _CAGR,
    _CAGR_div_retnStd,
    _CAGR_div_UI,
    SPY_retnStd_d_UI,     
    SPY_CAGR,
    SPY_CAGR_d_retnStd,
    SPY_CAGR_d_UI,
  ) = symb_perf_stats_vectorized_v8(df_SPY)  

  print(f'\nSPY: retnStd/UI, CAGR/retnStd, CAGR/UI, CAGR: {SPY_retnStd_d_UI[0]:>17,.1f}, {SPY_CAGR_d_retnStd[0]:>17,.1f}, {SPY_CAGR_d_UI[0]:>17.1f}, {SPY_CAGR[0]:>17.1f}')

  # drop last list, with frequency count 2 or less, in l_sym_freq_cnt from zip
  zip_cnt_n_syms = zip([6,5,4,3], l_sym_freq_cnt[:-1])  
  for item in zip_cnt_n_syms:
    sym_freq_cnt = item[0]
    syms = item[1]
    if syms:  # iterate ONLY if there are symbols in syms
      df_eval = df_train[start_eval:end_eval][syms]      

      if verbose:
        print(f'start_eval: {start_eval},  date: {d_end_train}')
        print(f'end_eval:   {end_eval},  date: {d_end_eval},  df_eval last date: {d_df_eval_end}')
        print(f'frequency count of symbol(s): {sym_freq_cnt}')      
        # print(f'\ndf_eval:\n{df_eval}\n')

        syms_n_SPY = syms + ['SPY']
        df_eval_n_SPY = df_train[start_eval:end_eval][syms_n_SPY]   
        print(f'\ndf_eval_n_SPY:\n{df_eval_n_SPY}\n')

      (
        _symbols,
        _period_yr,
        _retn,
        _DD,
        _UI,
        _MDD,
        _retnMean,
        _retnStd,
        _retnStd_div_UI,
        _CAGR,
        _CAGR_div_retnStd,
        _CAGR_div_UI,
        grp_retnStd_d_UI,     
        grp_CAGR,
        grp_CAGR_d_retnStd,
        grp_CAGR_d_UI,
      ) = symb_perf_stats_vectorized_v8(df_eval)  
      print(f'grp(retnStd/UI):   mean, std, mean/std: {grp_retnStd_d_UI[0]  :>17,.1f}, {grp_retnStd_d_UI[1]  :>17,.1f}, {grp_retnStd_d_UI[2]  :>17,.1f}')
      print(f'grp(CAGR/retnStd): mean, std, mean/std: {grp_CAGR_d_retnStd[0]:>17,.1f}, {grp_CAGR_d_retnStd[1]:>17,.1f}, {grp_CAGR_d_retnStd[2]:>17,.1f}')
      print(f'grp(CAGR/UI):      mean, std, mean/std: {grp_CAGR_d_UI[0]     :>17,.1f}, {grp_CAGR_d_UI[1]     :>17,.1f}, {grp_CAGR_d_UI[2]     :>17,.1f}')
      print(f'grp(CAGR):         mean, std, mean/std: {grp_CAGR[0]          :>17,.1f}, {grp_CAGR[1]          :>17,.1f}, {grp_CAGR[2]          :>17,.1f}')

      if store_results:  # record results to df
        row_add0      = [n_samples, str(days_lookbacks), days_eval, n_top_syms, syms_start, syms_end, sym_freq_cnt]
        row_add1      = [grp_CAGR[0],           grp_CAGR[1],           grp_CAGR[2]]
        row_add2      = [grp_CAGR_d_UI[0],      grp_CAGR_d_UI[1],      grp_CAGR_d_UI[2]]
        row_add3      = [grp_CAGR_d_retnStd[0], grp_CAGR_d_retnStd[1], grp_CAGR_d_retnStd[2]]
        row_add4      = [grp_retnStd_d_UI[0],   grp_retnStd_d_UI[1],   grp_retnStd_d_UI[2]]
        row_add5      = [SPY_CAGR[0], SPY_CAGR_d_UI[0], SPY_CAGR_d_retnStd[0], SPY_retnStd_d_UI[0]]
        row_add_total = row_add0 + row_add1 + row_add2 + row_add3 + row_add4 + row_add5
        print(f'row_add_total: {row_add_total}')


        # df_eval_sym_freq_results.loc[len(df_eval_sym_freq_results)] = row_add_total
        # print(f'appended row_add to df_eval_sym_freq_results:\n{row_add_total}\n')
        df_eval_sym_freq_results_test.loc[len(df_eval_sym_freq_results_test)] = row_add_total       
        print(f'appended row_add to df_eval_sym_freq_results_test:\n{row_add_total}\n')        
  
  
  print('='*50, '\n')  

z_grp_top_set_syms_n_freq:
1 of 57 max_lookback_slice
max_lookback_slice: (19, 79, 83)
max lookback dates: 2022-08-18, 2022-11-11, 2022-11-17
df_eval dates (inclusive): 2022-11-11 - 2022-11-16
top_set_syms_n_freq: [('AXON', 6), ('GILD', 6), ('GBX', 5), ('LOCO', 5), ('RMBS', 5), ('SLB', 5), ('STRL', 5), ('ASC', 4), ('ELF', 4), ('EME', 4), ('SANM', 4), ('ATEN', 3), ('CAT', 3), ('CCRN', 3), ('FTI', 3), ('HEES', 3), ('HOG', 3), ('NTNX', 3), ('SKT', 3), ('TWNK', 3)]

sym_freq_6: ['AXON', 'GILD']
sym_freq_5: ['GBX', 'LOCO', 'RMBS', 'SLB', 'STRL']
sym_freq_4: ['ASC', 'ELF', 'EME', 'SANM']
sym_freq_3: ['ATEN', 'CAT', 'CCRN', 'FTI', 'HEES', 'HOG', 'NTNX', 'SKT', 'TWNK']


SPY: retnStd/UI, CAGR/retnStd, CAGR/UI, CAGR:               1.4,             -60.9,             -83.2,              -0.5
grp(retnStd/UI):   mean, std, mean/std:               6.0,               2.3,               2.6
grp(CAGR/retnStd): mean, std, mean/std:           1,063.6,             738.8,               1.4
grp(CAGR/UI):  

In [482]:
# pickle_dump(df_eval_sym_freq_results, path_data_dump, fp_df_eval_sym_freq_results)
pickle_dump(df_eval_sym_freq_results_test, path_data_dump, fp_df_eval_sym_freq_results_test)

In [483]:
# https://towardsdatascience.com/pandas-groupby-a-simple-but-detailed-tutorial-314b8f37005d
# https://towardsdatascience.com/accessing-data-in-a-multiindex-dataframe-in-pandas-569e8767201d
# https://towardsdatascience.com/summarizing-data-with-pandas-crosstab-efc8b9abecf
# https://towardsdatascience.com/how-to-flatten-multiindex-columns-and-rows-in-pandas-f5406c50e569
# https://datascientyst.com/list-aggregation-functions-aggfunc-groupby-pandas/

In [494]:
df = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_test)
df

n_samples   days_lookbacks days_eval n_top_syms syms_start syms_end sym_freq_cnt  grp(CAGR)_mean  grp(CAGR)_std  grp(CAGR)_mean/std  grp(CAGR/UI)_mean  grp(CAGR/UI)_std  grp(CAGR/UI)_mean/std  grp(CAGR/retnStd)_mean  grp(CAGR/retnStd)_std  grp(CAGR/retnStd)_mean/std  grp(retnStd/UI)_mean  grp(retnStd/UI)_std  grp(retnStd/UI)_mean/std  SPY_CAGR  SPY_CAGR/UI  SPY_CAGR/retnStd  SPY_retnStd/UI
0           57  [15, 30, 60,...         6         20          0       10            6       94.562870      95.360736         0.991633        2.326079e+08       2.326080e+08          1.000000            7572.469282             7689.817721               0.984740                15240.906112          15240.497106             1.000027           1.274855   189.903417       103.706078         1.831170
1           57  [15, 30, 60,...         6         20          0       10            5        4.160346      10.496394         0.396360        4.257191e+02       1.006395e+03          0.423014             154.159131              405.007470               0.380633                    1.179793              0.629508             1.874152           1.274855   189.903417       103.706078         1.831170
2           57  [15, 30, 60,...         6         20          0       10            4        1.962740       4.378000         0.448319        6.963529e+03       1.857026e+04          0.374983             124.840658              202.122558               0.617648                   53.401984            145.992902             0.365785           1.274855   189.903417       103.706078         1.831170
3           57  [15, 30, 60,...         6         20          0       10            3    67966.293119   96120.121112         0.707097        3.447739e+06       4.875893e+06          0.707099          598687.587472           846739.482696               0.707051                    2.455574              2.338732             1.049960           1.274855   189.903417       103.706078         1.831170
4           57  [15, 30, 60,...         6         20          0       10            6       -0.468234       0.000000             -inf       -2.598420e+01       0.000000e+00              -inf             -41.701153                0.000000                   -inf                    0.623105              0.000000                  inf           1.911171   232.447231       181.632041         1.279770
...        ...              ...       ...        ...        ...      ...          ...             ...            ...              ...                 ...                ...               ...                    ...                     ...                    ...                         ...                   ...                  ...                ...          ...              ...              ...
3631        57     [15, 30, 60]         4         20          0       10            3      239.713284     681.979210         0.351496        9.964980e+04       2.886233e+05          0.345259            6576.759363            18269.100593               0.359994                    5.127704              5.371460             0.954620           9.290311  2215.846998       688.380743         3.218926
3632        57     [15, 30, 60]         4         20          0       10            6       -0.611405       0.388357        -1.574336       -1.463245e+01       2.935177e+00         -4.985204             -14.881916                4.466643              -3.331790                    1.015522              0.107567             9.440860           1.711979   251.067968       122.307583         2.052759
3633        57     [15, 30, 60]         4         20          0       10            5       97.643535     159.467868         0.612309        1.947960e+08       3.192322e+08          0.610202            6436.941038            10070.086910               0.639214                12103.413633          13116.478091             0.922764           1.711979   251.067968       122.307583         2.052759
3634        57     [15, 3

In [495]:
# pd.crosstab(df['days_lookbacks'], df['grp(CAGR/UI)_mean'])
tbl = df.groupby(['days_lookbacks', 'days_eval', 'sym_freq_cnt'])\
        .agg({'grp(CAGR)_mean':          ['count', 'mean', 'std',],
              'grp(CAGR/UI)_mean':       ['mean', 'std',],
              'grp(CAGR/retnStd)_mean':  ['mean', 'std',],
              'SPY_CAGR':                ['mean', 'std',],
              'SPY_CAGR/UI':             ['mean', 'std',],
              'SPY_CAGR/retnStd':        ['mean', 'std',],                           
              })

In [496]:
tbl

grp(CAGR)_mean                             grp(CAGR/UI)_mean               grp(CAGR/retnStd)_mean                SPY_CAGR              SPY_CAGR/UI               SPY_CAGR/retnStd             
                                                 count          mean           std              mean           std                   mean           std      mean        std          mean           std             mean          std
days_lookbacks   days_eval sym_freq_cnt                                                                                                                                                                                               
[120]            4         3                        57  7.160189e+04  3.787630e+05     1.431899e+11   7.575289e+11     1.219301e+06        6.443590e+06  1.622151   5.560079  1.888205e+06  9.981365e+06       583.339314  2648.752670
                 5         3                        57  4.097137e+03  1.173244e+04     1.703780e+09   5.008042e+09     6.767117e+04        1.950623e+05  1.352828   2.055144  1.382311e+03  2.798544e+03       156.500174   234.768493
                 6         3                        57  1.146583e+03  3.291551e+03     1.081330e+09   3.513702e+09     2.079432e+04        5.922263e+04  0.808830   1.630601  7.219396e+02  2.497623e+03        92.317989   204.897099
[15, 30, 60, ... 4         3                        51  8.299452e+08  3.347058e+09     1.657202e+15   6.694768e+15     7.203621e+09        2.905836e+10  1.886524   6.457245  3.498366e+06  1.274541e+07       893.248168  3376.069642
                           4                        57  1.167023e+04  3.529550e+04     6.611386e+09   2.345191e+10     2.311203e+05        7.189044e+05  1.596529   6.160858  3.130111e+06  1.209189e+07       788.537118  3204.919670
                           5                        57  8.300062e+01  3.854488e+02     4.563672e+07   1.870784e+08     4.053242e+03        1.322700e+04  1.596529   6.160858  3.130111e+06  1.209189e+07       788.537118  3204.919670
                           6                        54  1.749694e+01  3.648385e+01     1.877524e+07   5.369208e+07     1.076292e+03        2.258692e+03  1.700482   6.316284  3.304008e+06  1.240584e+07       833.829183  3288.343789
                 5         3                        54  1.642821e+06  6.758321e+06     3.049406e+08   1.202337e+09     1.449622e+07        5.959657e+07  0.870845   1.706578  1.115844e+03  2.596571e+03       106.165088   207.908369
                           4                        57  2.751256e+02  1.043525e+03     3.746738e+07   1.869170e+08     6.376060e+03        2.170587e+04  0.788441   1.697291  1.054953e+03  2.539473e+03        96.680240   206.297008
                           5                        57  1.253535e+01  1.994739e+01     9.240818e+05   6.893529e+06     3.754604e+02        5.440509e+02  0.788441   1.697291  1.054953e+03  2.539473e+03        96.680240   206.297008
                           6                        56  2.088053e+01  4.527283e+01     1.822342e+07   7.724905e+07     1.349062e+03        3.055960e+03  0.813356   1.702101  1.074763e+03  2.558008e+03        99.507361   207.046815
                 6         3                        57  3.774442e+04  2.444855e+05     6.881932e+06   4.773324e+07     3.595699e+05        2.351467e+06  0.804392   1.357303  6.541680e+02  2.030758e+03        96.026973   178.750138
                           4                        57  1.169252e+02  3.424817e+02     1.838596e+07   1.338837e+08     2.976231e+03        8.250728e+03  0.804392   1.357303  6.541680e+02  2.030758e+03        96.026973   178.750138
                           5                        57  5.669838e+00  9.232895e+00     1.789658e+06   1.350607e+07     2.061302e+02        4.955949e+02  0.804392   1.357303  6.541680e+02  2.030758e+03        96.026973   178.750138
                           6                        55  1.269940e+01  3.027095e+01     8.464103e+06   4.394287e+0

In [497]:
tbl['dif_CAGR_mean'] = tbl['grp(CAGR)_mean', 'mean'] - tbl['SPY_CAGR', 'mean']
tbl['dif_CAGR/UI_mean'] = tbl['grp(CAGR/UI)_mean', 'mean'] - tbl['SPY_CAGR/UI', 'mean'] 
tbl['dif_CAGR/retnStd_mean'] = tbl['grp(CAGR/retnStd)_mean', 'mean'] - tbl['SPY_CAGR/retnStd', 'mean'] 
tbl['grp_CAGR/UI_mean/std'] = tbl['grp(CAGR/UI)_mean', 'mean'] / tbl['grp(CAGR/UI)_mean', 'std']  
tbl['SPY_CAGR/UI_mean/std'] = tbl['SPY_CAGR/UI', 'mean'] / tbl['SPY_CAGR/UI', 'std'] 
tbl['dif_CAGR/UI_mean/std'] = tbl['grp_CAGR/UI_mean/std']  - tbl['SPY_CAGR/UI_mean/std'] 
# tbl.sort_values(by='dif_CAGR/UI_mean', ascending=False, inplace=True)
# tbl.sort_values(by='dif_CAGR/UI_mean/std', ascending=False, inplace=True)
# tbl

In [498]:
tbl.sort_values(by='dif_CAGR/UI_mean', ascending=False, inplace=True)
tbl.head(20)

grp(CAGR)_mean                             grp(CAGR/UI)_mean               grp(CAGR/retnStd)_mean                SPY_CAGR              SPY_CAGR/UI               SPY_CAGR/retnStd              dif_CAGR_mean dif_CAGR/UI_mean dif_CAGR/retnStd_mean grp_CAGR/UI_mean/std SPY_CAGR/UI_mean/std dif_CAGR/UI_mean/std
                                                 count          mean           std              mean           std                   mean           std      mean        std          mean           std             mean          std                                                                                                                    
days_lookbacks   days_eval sym_freq_cnt                                                                                                                                                                                                                                                                                                                   
[15, 30, 60, ... 4         3                        51  8.299452e+08  3.347058e+09     1.657202e+15   6.694768e+15     7.203621e+09        2.905836e+10  1.886524   6.457245  3.498366e+06  1.274541e+07       893.248168  3376.069642  8.299452e+08     1.657202e+15     7.203620e+09              0.247537             0.274480            -0.026944    
[30, 60, 120]    4         3                        57  1.674742e+08  1.243529e+09     3.295918e+14   2.487675e+15     1.510186e+09        1.109446e+10  1.166384   4.034191  1.019938e+06  7.134063e+06       350.485765  1890.882420  1.674742e+08     3.295918e+14     1.510186e+09              0.132490             0.142967            -0.010477    
[15, 30, 60]     4         3                        57  1.647513e+08  1.243844e+09     3.295023e+14   2.487687e+15     1.430325e+09        1.079868e+10  2.226557   5.238719  9.450168e+05  7.121750e+06       386.309895  1886.603151  1.647513e+08     3.295023e+14     1.430324e+09              0.132453             0.132694            -0.000241    
[120]            4         3                        57  7.160189e+04  3.787630e+05     1.431899e+11   7.575289e+11     1.219301e+06        6.443590e+06  1.622151   5.560079  1.888205e+06  9.981365e+06       583.339314  2648.752670  7.160027e+04     1.431880e+11     1.218717e+06              0.189022             0.189173            -0.000151    
[60, 120]        4         3                        57  4.692227e+04  1.912287e+05     8.952516e+10   3.830446e+11     7.948260e+05        3.253594e+06  3.390902   7.305688  4.000439e+06  1.382751e+07      1186.739437  3640.862631  4.691888e+04     8.952116e+10     7.936393e+05              0.233720             0.289310            -0.055590    
[15, 30]         4         5                        47  7.429784e+03  2.953478e+04     1.483294e+10   5.907553e+10     1.078809e+05        4.142346e+05  2.214770   5.570698  1.092209e+06  4.477893e+06       253.181806   691.286755  7.427569e+03     1.483185e+10     1.076277e+05              0.251084             0.243911             0.007173    
[30]             4         3                        57  4.659280e+03  1.680199e+04     9.203712e+09   3.362550e+10     7.501032e+04        2.414503e+05  5.090656  16.362122  3.687913e+06  2.097357e+07       600.468196  2141.941451  4.654190e+03     9.200024e+09     7.440985e+04              0.273712             0.175836             0.097876    
[15, 30]         4         4                        52  4.991486e+03  2.064628e+04     7.934713e+09   3.982220e+10     8.780744e+04        3.557989e+05  4.033014  14.185528  1.151816e+06  4.306616e+06       319.650886   810.660642  4.987453e+03     7.933561e+09     8.748779e+04              0.199254             0.267453            -0.068199    
[15, 30, 60, ... 4         4                        57  1.167023e+04  3.529550e+04     6.611386e+09   2.345191e+10     2.311203e+05        7.189044e+05  1.596529   6.160858  3.130111e+06  1.209189e+07       788.537118  3204

In [500]:
tbl.sort_values(by='dif_CAGR/UI_mean/std', ascending=False, inplace=True)
tbl.head(20)

grp(CAGR)_mean                              grp(CAGR/UI)_mean               grp(CAGR/retnStd)_mean                SPY_CAGR              SPY_CAGR/UI               SPY_CAGR/retnStd              dif_CAGR_mean dif_CAGR/UI_mean dif_CAGR/retnStd_mean grp_CAGR/UI_mean/std SPY_CAGR/UI_mean/std dif_CAGR/UI_mean/std
                                                 count          mean            std              mean           std                   mean           std      mean        std          mean           std             mean          std                                                                                                                    
days_lookbacks   days_eval sym_freq_cnt                                                                                                                                                                                                                                                                                                                    
[60, 120]        6         5                        18      0.050627       0.564624     8.621612e+02   8.592889e+02       101.763560        1.498597e+02  0.631909   1.528731  1.662109e+02  2.600452e+02        62.998013   159.667079     -0.581282     6.959503e+02        38.765548              1.003343             0.639161             0.364181    
[30, 60, 120]    4         4                        57  31963.663491  117358.946491     1.405087e+07   3.416053e+07    660148.026660        2.422087e+06  1.166384   4.034191  1.019938e+06  7.134063e+06       350.485765  1890.882420  31962.497108     1.303093e+07    659797.540896              0.411319             0.142967             0.268351    
[15, 30, 60]     5         6                        48     10.366657      24.477676     6.729801e+06   2.012184e+07       781.643258        1.575575e+03  4.259215  10.931028  2.116748e+05  1.452483e+06       292.542102   539.483953      6.107442     6.518126e+06       489.101156              0.334453             0.145733             0.188720    
[30]             5         3                        56   2576.571556    9203.186909     1.693362e+09   5.566190e+09     39564.285293        1.146623e+05  2.701514   8.057148  1.099939e+06  6.970166e+06       287.246972   814.836284   2573.870042     1.692262e+09     39277.038320              0.304223             0.157807             0.146416    
[15, 30, 60]     4         4                        57   3108.132785   20076.948368     4.104172e+08   1.550378e+09     70842.597345        4.149813e+05  2.226557   5.238719  9.450168e+05  7.121750e+06       386.309895  1886.603151   3105.906228     4.094722e+08     70456.287450              0.264721             0.132694             0.132026    
                 5         3                        57  18663.249900  110987.998826     3.690644e+07   1.402812e+08    154891.058279        9.172549e+05  3.544903  10.155424  1.782551e+05  1.332933e+06       241.333798   509.153098  18659.704998     3.672818e+07    154649.724482              0.263089             0.133731             0.129358    
                 4         6                        48     11.985674      31.352522     1.613420e+07   5.933497e+07       896.500898        2.124279e+03  2.455172   5.649227  1.121963e+06  7.760736e+06       435.343932  2053.703523      9.530502     1.501224e+07       461.156966              0.271917             0.144569             0.127348    
                 5         5                        56    472.797956    1769.952347     1.029801e+09   3.964523e+09     28065.500400        1.115894e+05  3.624157  10.229527  1.814389e+05  1.344777e+06       247.771220   511.414924    469.173798     1.029620e+09     27817.729181              0.259754             0.134921             0.124833    
[30]             4         3                        57   4659.280363   16801.986360     9.203712e+09   3.362550e+10     75010.320703        2.414503e+05  5.090656  16.362122  3.687913e+06  2.097357e+07       600.

In [490]:
if verbose:
  # from IPython.display import display, HTML
  # from myUtils import symb_perf_stats_vectorized_v8
  df = df_eval.copy()
  df_SPY = df_SPY.copy()  
  _my_df = [df_eval, df_SPY]
  for _df in _my_df:
    # print(f'_df:\n{_df}')
    (
      symbols,
      period_yr,
      retn,
      DD,
      UI,
      MDD,
      retnMean,
      retnStd,
      retnStd_div_UI,
      CAGR,
      CAGR_div_retnStd,
      CAGR_div_UI,
      grp_retnStd_div_UI,     
      grp_CAGR,
      grp_CAGR_div_retnStd,
      grp_CAGR_div_UI,
    ) = symb_perf_stats_vectorized_v8(_df)

    # display(HTML(_df.to_html()))
    print(f'_df:\n{_df}\n')
    print(f'symbols:\n{symbols}\n')
    print(f'period_yr:\n{period_yr}\n')
    print(f'retn:\n{retn}\n')
    print(f'DD:\n{DD}\n')
    print(f'UI:\n{UI}\n')
    print(f'MDD:\n{MDD}\n')
    print(f'retnMean:\n{retnMean}\n')
    print(f'retnStd:\n{retnStd}\n')
    print(f'retnStd_div_UI:\n{retnStd_div_UI}\n')
    print(f'CAGR:\n{CAGR}\n')
    print(f'CAGR_div_retnStd:\n{CAGR_div_retnStd}\n')
    print(f'CAGR_div_UI:\n{CAGR_div_UI}\n')
    print(f'grp_retnStd_div_UI:\n{grp_retnStd_div_UI}\n')
    print(f'grp_CAGR:\n{grp_CAGR}\n')
    print(f'grp_CAGR_div_retnStd:\n{grp_CAGR_div_retnStd}\n')
    print(f'grp_CAGR_div_UI:\n{grp_CAGR_div_UI}\n')            

In [491]:
# # https://stackoverflow.com/questions/31674195/plot-normal-distribution-given-mean-and-sigma-python
# # loc is mean, scale is standard deviation
# import pylab
# import numpy as np
# from scipy.stats import norm
# # x = np.linspace(-10000,100000,1000)
# x = np.linspace(-40e+10,50e+10,1000)
# y = norm.pdf(x, loc=2.562777e+10, scale=1.036925e+11)    # loc = mean, scale = standard deviation
# # z = norm.pdf(x, loc=3.540615e+10, scale=1.194430e+11)    # for example
# # z1 = norm.pdf(x, loc=298.805901, scale=826.875749)    # for example
# # z1 = norm.pdf(x, loc=1.021825, scale=1.505096)    # for example
# pylab.plot(x,y, 'b')
# # pylab.plot(x,z, 'g')
# # pylab.plot(x,z1, 'r')
# pylab.show()

In [492]:
# # get values of _cols, where grp(CAGR/retnStd)_mean is max after filtering out inf
# _cols = ['grp(CAGR/retnStd)_mean', 'grp(CAGR/retnStd)_std', 'grp(CAGR/retnStd)_mean/std']
# # _df_no_inf = df.loc[df['grp(CAGR/retnStd)_mean'] != np.inf]  # df with filter out inf in column grp(CAGR/UI)_mean 
# # _idx = _df_no_inf['grp(CAGR/retnStd)_mean'].idxmax()  # index value of max in grp(CAGR/UI)_mean 
# _idx = df['grp(CAGR/retnStd)_mean'].idxmax()  # index value of max in grp(CAGR/UI)_mean 
# grp_inf_replacement = df.loc[[_idx], _cols].squeeze()  # convert df (only has 1 row) to series
# print(f'_idx: {_idx}')
# grp_inf_replacement


In [493]:
# # get values of _cols, where SPY_CAGR/retnStd is max after filtering out inf
# _cols = ['SPY_CAGR/retnStd']
# # _df_no_inf = df.loc[df['SPY_CAGR/retnStd'] != np.inf]  # df with filter out inf in column grp(CAGR/UI)_mean 
# _idx = df['SPY_CAGR/retnStd'].idxmax()  # index value of max in grp(CAGR/UI)_mean 
# SPY_inf_replacement = df.loc[[_idx], _cols].squeeze()  # convert df (only has 1 row) to series
# print(f'_idx: {_idx}')
# SPY_inf_replacement